Add media player entity schemas

matus-tomlein opened the pull request:
Schemas for new context entities that will be attached to media player events to enable:

* tracking a session that groups all events coming from the same media player,
* tracking ads within media,
* tracking media player quality.

## `media_player_session` (#1266)

A session groups events from the same media player using a `sessionId` identifier (which may be provided by the user).

A secondary role of the session is a sort of edge analytics since it provides statistics about the playback (such as the time spent playing) in order to make data modelling more efficient.

## `media_player_ad_break` (#1268)

An ad break is a grouping of ads played back to back. The schema will identify all ad events belonging to the same ad break.

## `media_player_ad` (#1267)

Describes an ad being played at the time of the event. Provide metadata about the ad.

## `media_player_quality` (#1269)

Describes quality of experience for the media playback at a specific time. It will be tracked with `updateQuality` media player events so it will give a snapshot of the data quality rather than aggregated metrics for the whole playback.

greg-el commented on schemas/com.snowplowanalytics.snowplow/media_player_quality/jsonschema/1-0-0:
```diff
@@ -0,0 +1,58 @@
+        "number",
+        "null"
+      ],
+      "description": "The current bitrate in bps.",
```
Might be nice to use `Bits Per Second` rather than `bps`.

matus-tomlein replied:
Replaced `bps` with `bits per second`.

greg-el reviewed (commented):
A couple of things I picked up on from an initial skim over the schemas.

rlh1994 commented on schemas/com.snowplowanalytics.snowplow/media_player_ad/jsonschema/1-0-0:
```diff
@@ -0,0 +1,56 @@
+      "maxLength": 4096,
+      "description": "Unique identifier for the ad."
+    },
+    "podPosition": {
+      "type": [
+        "number",
+        "null"
+      ],
+      "description": "The position of the ad within the ad break, starting with 1.",
+      "minimum": 0,
+      "maximum": 65535
+    },
+    "duration": {
+      "type": [
+        "number",
+        "null"
+      ],
+      "description": "Length of the video ad in seconds.",
+      "minimum": 0,
+      "maximum": 9007199254740991
```
Could we be more explicit and make these integers?

matus-tomlein replied:
I wanted to make this have the same format as the `duration` property in the `media_player` schema with info about the main media content. That one allows for double precision values, so I thought we should do the same here.

rlh1994 replied:
Sorry I meant the `podPosition` instead?

rlh1994 commented on schemas/com.snowplowanalytics.snowplow/media_player_ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,57 @@
+    "position": {
+      "type": "number",
+      "description": "The position of the ad break within the content, starting with 1.",
+      "minimum": 0,
```
Also, what does this actually mean? As in this is the nth ad break? I'd probably rephrase the description in that case as it wasn't immediatly clear to me that's what this meant.

matus-tomlein replied:
Thinking more about this, I don't really have a good use case for the ad break position property... I took from another schema as it seemed useful to me but now I can't find a good reason for having it. We already have the `startTime` of the ad break so we should be able to order them. Also I added a `totalAdBreaks` property to the `media_player_session` so we can easily know how many ad breaks there were. I removed it from the schema.

rlh1994 commented on schemas/com.snowplowanalytics.snowplow/media_player_quality/jsonschema/1-0-0:
```diff
@@ -0,0 +1,58 @@
+        "number",
+        "null"
+      ],
+      "description": "The number of dropped frames in the stream.",
```
Maybe just specify if this is cumulative or since the last event?

matus-tomlein replied:
Specified since last report as that would be my preference (happy to change if there are other opinions).

rlh1994 commented on schemas/com.snowplowanalytics.snowplow/media_player_session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,84 @@
+      "format": "uuid",
+      "description": "An identifier for the session."
+    },
+    "startedAt": {
+      "type": [
+        "null",
+        "string"
+      ],
+      "description": "Date-time timestamp of when the session started.",
+      "format": "date-time"
```
Should this not be a timestamp type?

matus-tomlein replied:
I think a date-time is more user friendly. This info is the actual time of the start of the session, it doesn't relate to the playback time so I think it's better to represent it in a readable format.

rlh1994 commented on schemas/com.snowplowanalytics.snowplow/media_player_session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,84 @@
+      "description": "Date-time timestamp of when the session started.",
+      "format": "date-time"
+    },
+    "pingIterval": {
+      "type": [
+        "number",
+        "null"
+      ],
+      "description": "Interval (seconds) in which the ping events will be sent. Default (specified in the tracker media docs) is assumed if not specified.",
+      "minimum": 0,
+      "maximum": 65535
```
integer?

matus-tomlein replied:
Right, fixed.

rlh1994 commented on schemas/com.snowplowanalytics.snowplow/media_player_session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,84 @@
+      "minimum": 0,
+      "maximum": 2147483647
+    },
+    "totalAds": {
+      "type": [
+        "number",
+        "null"
+      ],
+      "description": "Number of ads played.",
+      "minimum": 0,
+      "maximum": 65535
```
Do we want to add the number of ad breaks as well? Also, integer?

matus-tomlein replied:
I think that makes sense, added `totalAdBreaks` and fixed integer type.

matus-tomlein commented on schemas/com.snowplowanalytics.snowplow/media_player_ad/jsonschema/1-0-0:
That's a very good point. The VAST schema (a standard for representing ads) has a `skipoffset` property which specifies the position when the ad skip controls appear, but I think that's too granular for us. So I opted for the `skippable` boolean.

matus-tomlein commented on schemas/com.snowplowanalytics.snowplow/media_player_ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,57 @@
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a context entity that is shared for all media player ad events that belong to the ad break",
```
Sounds better, changed!

matus-tomlein commented on schemas/com.snowplowanalytics.snowplow/media_player_quality/jsonschema/1-0-0:
```diff
@@ -0,0 +1,58 @@
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a context entity that is tracked with media player events on data quality change",
```
I think both cases – triggered by manual selection of quality by user but also by an automatic change in the player. We should have different event types for the two cases – maybe `updateQuality` and `userUpdateQuality`?

miike replied:
Agreed on separating. Automatic updates might get quite noisy whereas user ones are more likely to be useful.

matus-tomlein replied:
In PR #1283, I added two event schemas for the two types of events:

* `media_player_event_quality_change` – Media player event tracked when the video playback quality changes automatically.
* `media_player_event_user_update_quality` – Media player event tracked when the video playback quality changes as a result of user interaction (choosing a different quality setting).

miike commented on schemas/com.snowplowanalytics.snowplow/media_player_ping_event/jsonschema/1-0-0:
```diff
@@ -0,0 +1,23 @@
+  },
+  "type": "object",
+  "properties": {
+    "label": {
```
Might be nice if we can re-use this for percentage based progress as well. Do we want to have units somewhere?

matus-tomlein replied:
In PR #1283, I added a schema for the `percent_progress` events as well.

We store the current progress using the `currentTime` and `percentProgress` properties in the [`media_player` entity schema](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/media_player/jsonschema/1-0-0#L47).

Is that what you mean by units or did I misunderstand?

matus-tomlein replied:
We addressed this in the PR for events.

miike commented on schemas/com.snowplowanalytics.snowplow/media_player_ad_event/jsonschema/1-0-0:
```diff
@@ -0,0 +1,31 @@
+  },
+  "type": "object",
+  "properties": {
+    "type": {
```
I would go so far as to separate this out into different schemas entirely (e.g., ad_start, ad_end etc) so that we can capture the properties separately.

matus-tomlein replied:
Right, I separated out the individual event schemas into another PR – #1283
There are quite a few of them, let me know if there is something I can do to make it easier to review!

miike commented:
Adding @jethron for review as well.

matus-tomlein commented:
Update: I created a new PR #1283 where I added schemas for media player events.

jethron commented on schemas/com.snowplowanalytics.snowplow/media_player_ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,50 @@
+    },
+    "startTime": {
+      "type": "number",
+      "description": "Playback time at the start of the ad break.",
```
Can we specify a unit here? In context it's obviously seconds, but we also have `timeToStart` below that's milliseconds so probably good to be explicit.

matus-tomlein replied:
Good point, added units!

jethron commented on schemas/com.snowplowanalytics.snowplow/media_player_session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,102 @@
+        "number",
+        "null"
+      ],
+      "description": "Total seconds of the content played. Each part of the content played is counted once (i.e., counts rewinding or rewatching the same content only once). Playback rate does not affect this value.",
```
This is the only one that mentions playback rate, does that mean the other fields implicitly _are_ affected by playback rate?

matus-tomlein replied:
None of them are actually, but the other ones talk about the time user spent which I thought makes it more clear that playback rate can't affect it. On the other hand, "seconds of content played" sounds more unclear to me whether it can be affected by the playback rate. Happy to add the note to the other field descriptions if you think that would help.

matus-tomlein commented on schemas/com.snowplowanalytics.snowplow/media_player_ad/jsonschema/1-0-0:
```diff
@@ -0,0 +1,63 @@
+    },
+    "adId": {
+      "type": "string",
+      "format": "uuid",
```
That's a great point, UUID was too restrictive as the ID will be passed from users. Thinking about this, I also changed the `mediaSessionId` and `breakId` in the session and ad-break schemas, because those might also be passed by users and with different formats.

matus-tomlein commented:
Update about what has changed in the last force push:

### `media_player_session` 

* rename timePlayedContent to timePlayed
* rename timePlayedContentMuted to timePlayedMuted
* rename contentDurationPlayed to contentWatched
* add timePaused
* rename totalAds to ads
* rename totalAdBreaks to adBreaks
* add adsClicked
* add adsSkipped
* add timeBuffering

### `media_player_quality`

Decided not to create this schema because the data quality information provided by the players is different in each one of them and it’s very likely that users will need to define a custom entity anyway to fit the information provided by their player and the way that the player provides the information. Happy to reconsider this if we figure out the common properties, but I don't think that I have (every time I look at a different player, I can see that other info needs to be tracked).

Additionally, I added a `timeBuffering` property to the session entity which calculates the amount of time that the player was buffering – this is a universal property that can be tracked anywhere and also the tracker can track it automatically (based on the buffer start and end events).

miike commented on schemas/com.snowplowanalytics.snowplow/media_player_ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,50 @@
+      "type": "number",
+      "description": "Playback time in seconds at the start of the ad break.",
+      "minimum": 0,
+      "maximum": 9007199254740991
```
Why is this max so high?

matus-tomlein replied:
This was because it was based on the previous `media_player` schema which had these limits for playback time. With the new `player` schema, I have changed this now to 2147483647.

miike commented on schemas/com.snowplowanalytics.snowplow/media_player_session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,138 @@
+        "null"
+      ],
+      "description": "Average playback rate (1 is normal speed).",
+      "minimum": -9007199254740991,
```
I'm not sure if we need a negative playback rate (technically that be playing the video backwards I think?)

matus-tomlein replied:
This was also based on the older `media_player` schema that had this range. I changed it to the range 0 to 16 now (I found that 16 is usually the max rate on browsers).

matus-tomlein commented:
**Update:**

1. changed the vendor for all media entities to `com.snowplowanalytics.snowplow.media` and renamed `media_player_session` to `session` and similar for the ad schemas.
2. added a new `player` schema that should be an updated version of the [`media_player` schema](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/media_player/jsonschema/1-0-0).
    * added properties to it: `label` (used to be in events before), `playerType` and `mediaType` (used in the media dbt model but haven't had a single place before), `fullscreen`, `pictureInPicture`
    * made some properties optional: `muted`, `volume`, `playbackRate`, `loop`
    * removed the `percentProgress` property which is now reported in the percent progress event and I don't think it's that useful because the schema has `currentTime` and `duration` properties anyway
    * changed max values for `currentTime` and `duration` to 2147483647 (was too large before)
    * changed min and max for `playbackRate` to 0 to 16
3. `ad_break` schema – decreased max value for `startTime`
4. `ad` schema
    * removed `percentProgress` which is now reported in `quartile`, `skip` and `click` events
    * decreased max value for `duration`

matus-tomlein commented:
Update: added a `quality` attribute to the `player` schema.

miike commented on schemas/com.snowplowanalytics.snowplow.media/ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,50 @@
+    },
+    "breakType": {
+      "description": "Type of ads within the break: linear (take full control of the video for a period of time), nonlinear (run concurrently to the video), companion (accompany the video but placed outside the player).",
+      "enum": [
```
@jethron Thoughts on removing the enum here? There may be other things we otherwise need to include (overlays - complete or partial, outstream (not sure if we should include these at all?).

jethron replied:
IMO overlays is a special case of companion and outstream is still linear; the context of the player's location in the page isn't captured in this property so I don't think should be factored in here. It _might_ be useful for users to populate this freeform, but probably only worth the flexibility if we don't intend specific handling of the value in the modeling stage.

miike commented on schemas/com.snowplowanalytics.snowplow.media/session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,138 @@
+    "mediaSessionId": {
+      "type": "string",
+      "maxLength": 256,
+      "description": "An identifier for the media session."
```
We should describe if / when this is reset (or does that description belong in the tracker?)

matus-tomlein replied:
It is not automatically reset, but provided by the user for the media content played in the media player. I have clarified it here and also in the trackers.

miike commented on schemas/com.snowplowanalytics.snowplow.media/session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,138 @@
+        "null",
+        "string"
+      ],
+      "description": "Date-time timestamp of when the session started.",
```
utc or local time?

matus-tomlein replied:
I have used local time in the trackers – clarified it here as well.

miike commented on schemas/com.snowplowanalytics.snowplow.media/session/jsonschema/1-0-0:
```diff
@@ -0,0 +1,138 @@
+    },
+    "avgPlaybackRate": {
+      "type": [
+        "number",
```
Do we need to / plan to round this to a certain precision in the trackers?

jethron commented on schemas/com.snowplowanalytics.snowplow.media/ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,50 @@
+      "type": [
+        "string",
+        "null"
+      ]
```
Should there be like an optional `podSize` here to complement `podPosition` in `ad`?

matus-tomlein replied:
I think that is useful! Have added the `podSize` property to ad break schema.

jethron commented on schemas/com.snowplowanalytics.snowplow.media/ad/jsonschema/1-0-0:
```diff
@@ -0,0 +1,63 @@
+        "null"
+      ],
+      "description": "The position of the ad within the ad break, starting with 1.",
+      "minimum": 0,
```
This was probably covered, did we intend `minimum: 1`/`exclusiveMinimum: true` here to match the description?

matus-tomlein replied:
I do remember changing this already based on some comment, must have been in a different place... Have changed to minimum 1.

jethron commented on schemas/com.snowplowanalytics.snowplow.media/ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,50 @@
+  "type": "object",
+  "properties": {
+    "name": {
+      "description": "Ad break name (e.g., pre-roll, mid-roll, and post-roll).",
```
Nit: Trailing period in descriptions seems inconsistent between the various schemas.

matus-tomlein replied:
Unified on adding the period to all descriptions.

jethron commented on schemas/com.snowplowanalytics.snowplow.media/player/jsonschema/1-0-0:
```diff
@@ -0,0 +1,131 @@
+  "properties": {
+    "currentTime": {
+      "type": "number",
+      "description": "The current playback time",
```
`in seconds`? Should this mention "position" at all or is that more confusing (thinking `player.currentTime` sounds a lot like `sum(session.timePlayed)`)?

matus-tomlein replied:
Good point, the property is based on the [HTML5 player one](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/currentTime) but I agree that is a bit confusing considering the stats in session. Have changed the description to clarify it more and mention position.

jethron commented on schemas/com.snowplowanalytics.snowplow.media/player/jsonschema/1-0-0:
```diff
@@ -0,0 +1,131 @@
+        "number",
+        "null"
+      ],
+      "description": "A double-precision floating-point value indicating the duration of the media in seconds",
```
`double-precision` seems like an unnecessary implementation detail.

matus-tomlein replied:
Removed the double-precision part.

jethron commented on schemas/com.snowplowanalytics.snowplow.media/player/jsonschema/1-0-0:
```diff
@@ -0,0 +1,131 @@
+      ],
+      "description": "Whether the video element is fullscreen"
+    },
+    "isLive": {
```
Maybe drop `is` to be consistent with the other flags? Maybe `livestream` or something?

matus-tomlein replied:
True, have changed to `livestream` to be consistent with other boolean properties.

jethron commented on schemas/com.snowplowanalytics.snowplow.media/ad_break/jsonschema/1-0-0:
```diff
@@ -0,0 +1,50 @@
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a context entity, shared with all media_player_ad events belonging to the ad break.",
```
Not sure how to reference the new name unambiguously. :sweat_smile: `com.snowplowanalytics.snowplow.media/ad` is a bit more of a mouthful.

matus-tomlein replied:
The `media_player_ad` was a name of what I originally planned to be a single schema for all ad events but we have diverged from that so just changed it to "ad events".

jethron reviewed (approved):
LGTM!

matus-tomlein commented:
Integrated into R142 PR #1302
